src/main/java/games/strategy/engine/framework/ServerGame.java

```diff
@@ -416,46 +395,59 @@ private void runStep(final boolean stepIsRestoredFromSavedGame) {
     if (m_isGameOver) {
       return;
     }
+    final GameStep currentStep = m_data.getSequence().getStep();
+    final IDelegate currentDelegate = currentStep.getDelegate();
+    if (!stepIsRestoredFromSavedGame
+        && currentDelegate.getClass().isAnnotationPresent(AutoSave.class)
+        && currentDelegate.getClass().getAnnotation(AutoSave.class).beforeStepStart()) {
+      autoSaveBefore(currentDelegate);
+    }
     startStep(stepIsRestoredFromSavedGame);
+    if (!stepIsRestoredFromSavedGame
+        && currentDelegate.getClass().isAnnotationPresent(AutoSave.class)
+        && currentDelegate.getClass().getAnnotation(AutoSave.class).afterStepStart()) {
+      autoSaveBefore(currentDelegate);
+    }
     if (m_isGameOver) {
       return;
     }
     waitForPlayerToFinishStep();
     if (m_isGameOver) {
       return;
     }
-    final boolean autoSaveAfterDelegateDone = endStep();
+    // save after the step has advanced
+    // otherwise, the delegate will execute again.
+    final boolean autoSaveThisDelegate = currentDelegate.getClass().isAnnotationPresent(AutoSave.class)
+        && currentDelegate.getClass().getAnnotation(AutoSave.class).afterStepEnd();
+    if (autoSaveThisDelegate && currentDelegate instanceof MoveDelegate) {
+      autoSave("autosaveAfter" + currentStep.getName().substring(0,1).toUpperCase()
+        + currentStep.getName().substring(1) + ".tsvg");
+    }
+    endStep();
     if (m_isGameOver) {
       return;
     }
     if (m_data.getSequence().next()) {
       m_data.getHistory().getHistoryWriter().startNextRound(m_data.getSequence().getRound());
-      autoSaveRound();
+      autoSave(m_data.getSequence().getRound() % 2 == 0
+          ? SaveGameFileChooser.getAutoSaveEvenFileName() : SaveGameFileChooser.getAutoSaveOddFileName());
     }
-    // save after the step has advanced
-    // otherwise, the delegate will execute again.
-    if (autoSaveAfterDelegateDone) {
-      autoSave();
+    if (autoSaveThisDelegate && !(currentDelegate instanceof MoveDelegate)) {
```

Comment:
There are some aspects in this code section that I like a good bit. Though, looking at a couple problems that still remain, like the `instanceof` check for example, it does seem that moving logic to the delegate object would simplify much of this. We could ask the delegate directly for what the auto-save name should be. It might even make sense to make the autosave functionality more direct and built into the delegate. To illustrate, I think this code could turn into:
```

GamePhase phase = gameData.getCurrentGamePhase();
if(delegate.shouldAutoSave(phase)) {
  autoSave(delegate.getAutoSaveName());
}
```

Autosave has enough to it, I don't think GameServer should know the details of how to autosave. Extracting this to an autosave module seems appropriate to me.

Regardless, moving logic to the delegate is pretty key. It simplifies, and let's you test most of this code relatively easily. Set up a delegate, then assert that given an input, it tells you whether you should autosave, and which file name.

What are your thoughts on this @simon33-2 , do you agree it would help test-ability?